Better exception when accessing two non-XA datasources in a transaction

### Description

#39283 revealed that Agroal 2.3+ now makes it possible to detect that an application is accessing two non-XA datasources in a transaction, which is unsafe.

When detected, we get an exception looking like this:

```
2024-05-02 11:55:18,201 WARN  [org.apa.cam.com.sch.SchedulerConsumer] (Camel (camel-1) thread #1 - scheduler://read-clean-write) Error processing exchange. Exchange[2388535E05D1BE3-0000000000000000]. Caused by: [org.springframework.jdbc.CannotGetJdbcConnectionException - Failed to obtain JDBC Connection]: org.springframework.jdbc.CannotGetJdbcConnectionException: Failed to obtain JDBC Connection
        at org.springframework.jdbc.datasource.DataSourceUtils.getConnection(DataSourceUtils.java:84)
        at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:653)
        at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:695)
        at org.apache.camel.component.sql.SqlProducer.processInternal(SqlProducer.java:150)
        at org.apache.camel.component.sql.SqlProducer.process(SqlProducer.java:137)
        at org.apache.camel.support.AsyncProcessorConverterHelper$ProcessorToAsyncProcessorBridge.process(AsyncProcessorConverterHelper.java:65)
        at org.apache.camel.processor.SendProcessor.process(SendProcessor.java:210)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.handleFirst(RedeliveryErrorHandler.java:462)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.run(RedeliveryErrorHandler.java:438)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.executeFromQueue(DefaultReactiveExecutor.java:240)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.executeFromQueue(DefaultReactiveExecutor.java:77)
        at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.await(DefaultAsyncProcessorAwaitManager.java:95)
        at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.process(DefaultAsyncProcessorAwaitManager.java:84)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:200)
        at org.apache.camel.impl.engine.CamelInternalProcessor.processTransacted(CamelInternalProcessor.java:397)
        at org.apache.camel.impl.engine.CamelInternalProcessor.process(CamelInternalProcessor.java:327)
        at org.apache.camel.processor.Pipeline$PipelineTask.run(Pipeline.java:102)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.executeFromQueue(DefaultReactiveExecutor.java:240)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.executeFromQueue(DefaultReactiveExecutor.java:77)
        at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.await(DefaultAsyncProcessorAwaitManager.java:95)
        at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.process(DefaultAsyncProcessorAwaitManager.java:84)
        at org.apache.camel.support.AsyncProcessorSupport.process(AsyncProcessorSupport.java:32)
        at org.apache.camel.impl.engine.CamelInternalProcessor.processTransacted(CamelInternalProcessor.java:397)
        at org.apache.camel.impl.engine.CamelInternalProcessor.process(CamelInternalProcessor.java:327)
        at org.apache.camel.processor.Pipeline$PipelineTask.run(Pipeline.java:102)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.executeFromQueue(DefaultReactiveExecutor.java:240)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.executeFromQueue(DefaultReactiveExecutor.java:77)
        at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.await(DefaultAsyncProcessorAwaitManager.java:95)
        at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.process(DefaultAsyncProcessorAwaitManager.java:84)
        at org.apache.camel.support.AsyncProcessorSupport.process(AsyncProcessorSupport.java:32)
        at org.apache.camel.jta.TransactionErrorHandler.processByErrorHandler(TransactionErrorHandler.java:234)
        at org.apache.camel.jta.TransactionErrorHandler$1.run(TransactionErrorHandler.java:197)
        at org.apache.camel.quarkus.component.jta.TransactionalJtaTransactionPolicy.runWithTransaction(TransactionalJtaTransactionPolicy.java:47)
        at org.apache.camel.quarkus.component.jta.RequiredJtaTransactionPolicy.run(RequiredJtaTransactionPolicy.java:26)
        at org.apache.camel.jta.TransactionErrorHandler.doInTransactionTemplate(TransactionErrorHandler.java:187)
        at org.apache.camel.jta.TransactionErrorHandler.processInTransaction(TransactionErrorHandler.java:138)
        at org.apache.camel.jta.TransactionErrorHandler.process(TransactionErrorHandler.java:102)
        at org.apache.camel.jta.TransactionErrorHandler.process(TransactionErrorHandler.java:111)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.handleFirst(RedeliveryErrorHandler.java:462)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.run(RedeliveryErrorHandler.java:438)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.doRun(DefaultReactiveExecutor.java:199)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.executeReactiveWork(DefaultReactiveExecutor.java:189)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.tryExecuteReactiveWork(DefaultReactiveExecutor.java:166)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:148)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:59)
        at org.apache.camel.processor.Pipeline.process(Pipeline.java:163)
        at org.apache.camel.impl.engine.CamelInternalProcessor.processNonTransacted(CamelInternalProcessor.java:354)
        at org.apache.camel.impl.engine.CamelInternalProcessor.process(CamelInternalProcessor.java:330)
        at org.apache.camel.component.scheduler.SchedulerConsumer.sendTimerExchange(SchedulerConsumer.java:70)
        at org.apache.camel.component.scheduler.SchedulerConsumer.poll(SchedulerConsumer.java:50)
        at org.apache.camel.support.ScheduledPollConsumer.doRun(ScheduledPollConsumer.java:204)
        at org.apache.camel.support.ScheduledPollConsumer.run(ScheduledPollConsumer.java:118)
        at java.base@21.0.2/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base@21.0.2/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
        at java.base@21.0.2/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base@21.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base@21.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base@21.0.2/java.lang.Thread.runWith(Thread.java:1596)
        at java.base@21.0.2/java.lang.Thread.run(Thread.java:1583)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:833)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
Caused by: java.sql.SQLException: Exception in association of connection to existing transaction
        at io.agroal.narayana.NarayanaTransactionIntegration.associate(NarayanaTransactionIntegration.java:130)
        at io.agroal.pool.ConnectionPool.getConnection(ConnectionPool.java:257)
        at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
        at io.agroal.api.AgroalDataSource_syYGL47zwCt2LQzSIsPp-v9wXhg_Synthetic_ClientProxy.getConnection(Unknown Source)
        at org.springframework.jdbc.datasource.DataSourceUtils.fetchConnection(DataSourceUtils.java:160)
        at org.springframework.jdbc.datasource.DataSourceUtils.doGetConnection(DataSourceUtils.java:118)
        at org.springframework.jdbc.datasource.DataSourceUtils.getConnection(DataSourceUtils.java:81)
        ... 60 more
Caused by: java.sql.SQLException: Unable to enlist connection to existing transaction
        at io.agroal.narayana.NarayanaTransactionIntegration.associate(NarayanaTransactionIntegration.java:121)
        ... 66 more
```

The root cause is really not helpful, because it doesn't specify *why* the connection couldn't be enlisted to the transaction. We should improve that message, and that means changes in Agroal.

@barreiro I think you had plans in that area; could you please link the corresponding Agroal Jira?

### Implementation ideas

Unfortunately `Transaction.enlistResource` only returns a boolean to state whether enlistment worked or not.

So here are our only solutions:

1. Rely on Narayana-specific interfaces that would expose an alternate version of `enlistResource` which throws an exception on failure, with details regarding what failed exactly ("Can't add two non-XA resources")
2. Just adapt the message in Agroal to provide a few common reasons for the failure, without certainty.

> Caused by: jakarta.transaction.RollbackException: ARJUNA016053: Could not commit transaction.

If you look at the cause of the exception (e.getCause()) then setRollbackOnly will appear in the backtrace.

> The stacktrace. What part of the user code is triggering the issue.
The narayana internal frames won't mean anything to them. Their own code's frames are there anyhow in the exception that agroal already throws.


> Caused by: jakarta.transaction.RollbackException: ARJUNA016053: Could not commit transaction.
        at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1283)
        at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
> Caused by: org.hibernate.exception.GenericJDBCException: Unable to acquire JDBC Connection [Exception in association of connection to existing transaction] [n/a]
        at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:63)
  
That actually is the deferred exception support at work, just in a different way. When commit is called, the transaction system MUST complete termination of the transaction before returning control. That is, it can't stop half way through terminating the tx and throw an exception. So if there is a problem, such as the pre-commit flush failing as here, then it catches and stashes that exception, goes on to finish the job by rolling back, then attaches the exception to the RollbackException it throws at the end. The same pattern can be used when a user explicitly calls setRollbackOnly rather than the tx system catching an exception - the stack at the point where the user or library called setRollbackOnly is stashed and thrown as cause if the user later tries to call commit. We could do that internally - catch the hibernate exception and call setRollbackOnly, relying on it to do the stack capture, but then the cause would rooted in the exception handler, not in hibernate.

Nothing would stop us stashing some more context when we internally call setRollbackOnly before returning false from enlistResource, but that would help the tx rollback message, not the one that's thrown by agroal. A slightly kludgy workaround would be to have enlistResource stash the offending LastResource instances on the tx context, from which agroal could retrieve them by tsr.getResource.

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -109,7 +109,7 @@
         <!-- When updating, align hibernate-search.version-for-documentation in docs/pom.xml -->
         <hibernate-search.version>7.1.1.Final</hibernate-search.version>
         <narayana.version>7.0.1.Final</narayana.version>
-        <agroal.version>2.3</agroal.version>
+        <agroal.version>2.4</agroal.version>
         <jboss-transaction-spi.version>8.0.0.Final</jboss-transaction-spi.version>
         <elasticsearch-opensource-components.version>8.13.4</elasticsearch-opensource-components.version>
         <rxjava.version>2.2.21</rxjava.version>

